consumer/pubsub/src/main/java/com/spotify/heroic/consumer/pubsub/PubSubConsumerModule.java

```diff
@@ -0,0 +1,321 @@
+        @PubSubScope
+        String topicId() { return topicId; }
+
+        @Provides
+        @Named("projectId")
+        @PubSubScope
+        String projectId() { return projectId; }
+
+        @Provides
+        @Named("maxOutstandingElementCount")
+        @PubSubScope
+        Long maxOutstandingElementCount() { return maxOutstandingElementCount; }
+
+        @Provides
+        @Named("maxOutstandingRequestBytes")
+        @PubSubScope
+        Long maxOutstandingRequestBytes() { return maxOutstandingRequestBytes; }
+
+        @Provides
+        @PubSubScope
+        ConsumerSchema.Consumer consumer(final IngestionManager ingestionManager) {
+            final IngestionGroup ingestion = ingestionManager.useDefaultGroup();
+
+            if (ingestion.isEmpty()) {
+                throw new IllegalStateException("No backends are part of the ingestion group");
+            }
+
+            final ConsumerSchema.Depends dep = DaggerConsumerSchema_Depends
+                .builder()
+                .primaryComponent(primary)
+                .depends(depends)
+                .dependsModule(new ConsumerSchema.DependsModule(ingestion))
+                .build();
+            final ConsumerSchema.Exposed exposed = schema.setup(dep);
+            return exposed.consumer();
+        }
+
+        @Provides
+        @PubSubScope
+        public Managed<Connection> connection(
+            final AsyncFramework async,
+            final ConsumerReporter reporter,
+            final ConsumerSchema.Consumer consumer,
+            @Named("maxOutstandingElementCount") Long maxOutstandingElementCount,
+            @Named("maxOutstandingRequestBytes") Long maxOutstandingRequestBytes,
+            @Named("consuming") AtomicInteger consuming,
+            @Named("total") AtomicInteger total,
+            @Named("errors") AtomicLong errors,
+            @Named("consumed") LongAdder consumed
+        ) {
+            return async.managed(new ManagedSetup<Connection>() {
+                @Override
+                public AsyncFuture<Connection> construct() {
+                    return async.call(() -> {
+                        log.info("p:{}, t:{}, s:{}", projectId, topicId, subscriptionId);
+                        final Connection connection = new Connection(
+                            consumer, projectId, topicId, subscriptionId, threads,
+                            maxOutstandingElementCount, maxOutstandingRequestBytes);
+                        connection.setEmulatorOptions();
+
+                        // Create topics/subscriptions if they don't exist
+                        connection.createTopic();
+                        connection.createSubscription();
+
+                        return connection.start();
+                    });
+                }
+
+                @Override
+                public AsyncFuture<Void> destruct(final Connection connection) {
+                    connection.shutdown();
+                    total.set(0);
```

Comment:
Do we need to reset any of the other values? Consuming, errors, consumed, etc

Reply:
I haven't read through the code that uses those so I'm not completely certain, but the kafka consumer just sets total to 0 in its destruct method.